src/test/emulators/workQueue.spec.ts

```diff
@@ -0,0 +1,114 @@
+import { expect } from "chai";
+
+import { WorkQueue } from "../../emulator/workQueue";
+import { FunctionsExecutionMode } from "../../emulator/types";
+
+function resolveIn(ms: number) {
+  if (ms === 0) {
+    return Promise.resolve();
+  }
+
+  return new Promise((res, rej) => {
+    setTimeout(res, ms);
+  });
+}
+
+describe("WorkQueue", () => {
+  describe("mode=AUTO", () => {
+    let queue: WorkQueue;
+    beforeEach(() => {
+      if (queue) {
+        queue.stop();
+      }
+      queue = new WorkQueue(FunctionsExecutionMode.AUTO);
+      queue.start();
+    });
+
+    it("runs a job immediately", () => {
+      let hasRun = false;
+      const work = () => {
+        hasRun = true;
+        return Promise.resolve();
+      };
+
+      queue.submit(work);
+      expect(hasRun, "hasRun");
```

Comment:
At lot of these tests, including this one, will fail if `work` is started async, even if immediately async. I think this can easily break seemingly harmless refactors, so I'd suggest adding `await resolveIn(100);` before this. Also, every time we mix sync and async callbacks (tasks in this case) we are at risk to [unleash Zalgo](https://blog.izs.me/2013/08/designing-apis-for-asynchrony) so I'd highly suggest that we modify the `WorkQueue` contract to ONLY call work async and test for it. i.e.

```js
queue.submit(work);
expect(!hasRun, "queue must not run work sync");
await resolveIn(100);
expect(hasRun, "hasRun async");
```

Reply:
Ok I modified this behavior a little bit so that it is:
  * In `AUTO` mode the work always runs right away
  * In `SEQUENTIAL` mode the work runs right away if the queue is empty, otherwise placed at the end of the queue.

Does that work?

Reply:
If by "right away" you mean "immediately at next tick but not this tick" then I'm for it. (i.e. it should pass the draft test that I posted in the previous comment.) In other words, I want this:

```js
var flag = false;
queue.submit(() => { flag = true; });
console.log(flag); // flag should be still false here, regardless of mode or queue-fullness
```

See the Zalgo article for the principle behind this -- but in short, if a callback MAY be called on a different tick, then please please make sure it is ALWAYS called on a different tick and NEVER called on the same tick.

Reply:
Ok this is now tested.